File: quickstart_guides/trees/java/InorderTraversal.java
/*
 *   Title: In-order traversal of binary tree
 *
 * Problem:
 *   Given a binary tree, return the inorder traversal of its nodes' values.
 *
 *   Execution: javac InorderTraversal.java && java InorderTraversal
 */
import java.util.*;


class InorderTraversal {

    public static class TreeNode {
     int val;
     TreeNode left;
     TreeNode right;
     TreeNode() {}
     TreeNode(int val) { this.val = val; }
     TreeNode(int val, TreeNode left, TreeNode right) {
         this.val = val;
         this.left = left;
         this.right = right;
     }
   }

   public List < Integer > inorderTraversal(TreeNode root) {
        List < Integer > res = new ArrayList < > ();
        Stack < TreeNode > stack = new Stack < > ();
        TreeNode curr = root;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            res.add(curr.val);
            curr = curr.right;
        }
        return res;
    }
    public static void main(String args[])
    {
        InorderTraversal it = new InorderTraversal();
        TreeNode tree = new TreeNode(1);
        tree.right = new TreeNode(2);
        tree.right.left = new TreeNode(3);

        System.out.println(it.inorderTraversal(tree));

        System.out.println("Passed all test cases");
    }
}

File: quickstart_guides/trees/python/inorder_traversal.py
"""
Title: In-order traversal

Problem:
    Given a binary tree, return the inorder traversal of its nodes' values.

Execution: python inorder_traversal.py
"""
from typing import List
import unittest


class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def inorder_traversal(root: TreeNode) -> List[int]:
    """Perform in-order traversal."""

    def traverse(root, out, visited):
        if root:
            out = traverse(root.left, out, visited)
            if root not in visited:
                visited.append(root)
                out.append(root.val)
            out = traverse(root.right, out, visited)
        return out

    return traverse(root, [], [])


class TestInorderTraversal(unittest.TestCase):
    """Unit tests for inorder_traversal."""

    def test_1(self):
        tn = TreeNode(1)
        tn.right = TreeNode(2)
        tn.right.left = TreeNode(3)
        self.assertEqual(inorder_traversal(tn), [1, 3, 2])


if __name__ == "__main__":
    unittest.main()
